app/src/main/java/com/arctouch/codechallenge/ui/home/HomeActivity.kt

```diff
@@ -0,0 +1,104 @@
+import android.os.PersistableBundle
+import android.support.v7.app.AppCompatActivity
+import android.support.v7.widget.LinearLayoutManager
+import android.support.v7.widget.RecyclerView
+import android.util.Log
+import android.view.View
+import com.arctouch.codechallenge.R
+import com.arctouch.codechallenge.model.Movie
+import com.arctouch.codechallenge.util.Logger
+import io.reactivex.Scheduler
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.schedulers.Schedulers
+import kotlinx.android.synthetic.main.home_activity.*
+import org.koin.android.viewmodel.ext.android.viewModel
+
+class HomeActivity : AppCompatActivity() {
+    private var TAG = Logger.tag
+    private val viewModel by viewModel<HomeViewModel>()
+
+    private val adapter: HomeAdapter by lazy {
+        HomeAdapter()
+    }
+
+    private var recyclerState: Parcelable? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        if (Logger.DEBUG) Log.d(TAG, "onCreate")
+
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.home_activity)
+        setupInterface()
+        loadData()
+    }
+
+    override fun onDestroy() {
+        if (Logger.DEBUG) Log.d(TAG, "onDestroy")
+
+        super.onDestroy()
+        viewModel.reset()
+    }
+
+    override fun onSaveInstanceState(outState: Bundle?, outPersistentState: PersistableBundle?) {
+        if (Logger.DEBUG) Log.d(TAG, "onSaveInstanceState")
+
+        super.onSaveInstanceState(outState, outPersistentState)
+        outState?.putParcelable("llmState", recyclerView.layoutManager?.onSaveInstanceState())
+    }
+
+    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
+        if (Logger.DEBUG) Log.d(TAG, "onRestoreInstanceState")
+
+        super.onRestoreInstanceState(savedInstanceState)
+        recyclerState = savedInstanceState?.getParcelable("llmState")
+    }
+
+    private fun setupInterface() {
+        if (Logger.DEBUG) Log.d(TAG, "setupInterface")
+
+        val llm = LinearLayoutManager(this)
+        recyclerView.layoutManager = llm
+        recyclerView.adapter = adapter
+        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
+            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
+                super.onScrollStateChanged(recyclerView, newState)
+                val lastVisibleItemPosition = llm.findLastVisibleItemPosition()
+                if (lastVisibleItemPosition == adapter.itemCount - 1 && !viewModel.isLoading) {
+                    loadData(viewModel.currentPage + 1)
+                }
+            }
+        })
+    }
+
+    @SuppressLint("CheckResult")
+    private fun loadData(page: Long = 1) {
+        if (Logger.DEBUG) Log.d(TAG, "loadData")
+
+        viewModel.requestMovies(page)
+                ?.subscribeOn(Schedulers.io())
+                ?.observeOn(AndroidSchedulers.mainThread())
+                ?.subscribe({ movie ->
```

Comment:
What would be your strategy to unit test how the app behaves depending on the API's result (success or error)? Can you foresee any challenges with your current implementation?

Reply:
In that I would create a loop to try to search multi-page movies. Taking into account that I know the number of pageings would have. So I would find out if all the pages are bringing a certain number of movies and if any page is giving error. And in the end I could test with a number in addition to see the error handling. As I'm using viewmodel and Koin, I could inject the ViewModel's dependency on the test and use its methods at will, I could also use RoboEletric for this, which I personally prefer.